### PR TITLE
Artifact unloading [ECR-4282] [ECR-4284]

### PR DESCRIPTION
**Pull request by slowli**
## Overview

This PR introduces artifact unloading into the core and Rust runtime. Internal docs on this subject can be found [here](https://wiki.bf.local/display/EXN/Unloading+Service+Artifacts).

Not done yet:

- [x] Document artifact unloading in the `runtime` module docs.

---

See: https://jira.bf.local/browse/ECR-4282, https://jira.bf.local/browse/ECR-4284

### CODE REVIEW

File: exonum/src/runtime/mod.rs [hunk 1]
```diff
@@ -329,11 +333,21 @@ impl fmt::Display for RuntimeFeature {
...
+/// GENESIS ::=
+///     deploy_artifact*
+///     (initiate_adding_service update_service_status)*
+///     after_commit
+/// RESUME ::= (deploy_artifact | update_service_status | migrate)* on_resume
 /// BLOCK* ::= PROPOSAL+ COMMIT
-/// PROPOSAL ::= before_transactions* (execute | initiate_adding_service)* after_transactions*
-/// COMMIT ::= deploy_artifact* update_service_status* after_commit
+/// PROPOSAL ::=
+///     (before_transactions CALL*)*
```
Comment (dmitry-timofeev) [inline exonum/src/runtime/mod.rs]: Are you saying "before_transactions can also execute (normal) transactions"?

File: exonum/src/runtime/types.rs [hunk 2]
```diff
@@ -470,20 +473,23 @@ impl ProtobufConvert for ArtifactStatus {
...
 
     fn to_pb(&self) -> Self::ProtoStruct {
+        use self::schema::lifecycle::ArtifactState_Status::*;
+
         match self {
-            ArtifactStatus::Active => schema::lifecycle::ArtifactState_Status::ACTIVE,
-            ArtifactStatus::Pending => schema::lifecycle::ArtifactState_Status::PENDING,
+            ArtifactStatus::Active => ACTIVE,
+            ArtifactStatus::Pending => PENDING,
+            ArtifactStatus::Unloading => NONE,
```
Comment (dmitry-timofeev) [inline exonum/src/runtime/types.rs]: Why isn't `unloading` added to the protobuf message?

File: exonum/src/runtime/types.rs [hunk 3]
```diff
@@ -470,20 +473,23 @@ impl ProtobufConvert for ArtifactStatus {
...
+
         Ok(match pb {
-            schema::lifecycle::ArtifactState_Status::ACTIVE => ArtifactStatus::Active,
-            schema::lifecycle::ArtifactState_Status::PENDING => ArtifactStatus::Pending,
-            schema::lifecycle::ArtifactState_Status::NONE => {
-                bail!("Status `NONE` is reserved for the further usage.")
-            }
+            ACTIVE => ArtifactStatus::Active,
+            PENDING => ArtifactStatus::Pending,
+            NONE => ArtifactStatus::Unloading,
```
Comment (dmitry-timofeev) [inline exonum/src/runtime/types.rs]: Is that correct? Either a new variant shall be added to the message, or NONE renamed to 'unloading'.

File: exonum/src/runtime/mod.rs (hunk 1 above)
Comment (slowli) [reply exonum/src/runtime/mod.rs]: Yeah, `before_transactions` and `after_transactions` handlers can execute calls to the service. I'll elaborate this below in the docs.

File: exonum/src/runtime/types.rs (hunk 3 above)
Comment (slowli) [reply exonum/src/runtime/types.rs]: I've renamed `NONE` -> `UNLOADING`.

Comment (dmitry-timofeev) [reply exonum/src/runtime/types.rs]: Looks good, thanks.

File: exonum/src/proto/schema/exonum/runtime/lifecycle.proto [hunk 4]
```diff
@@ -47,8 +47,8 @@ message GenesisConfig {
 message ArtifactState {
   // Status of an artifact deployment.
   enum Status {
-    // Reserved for the further usage.
-    NONE = 0;
+    // The artifact is pending unload.
+    UNLOADING = 0;
     // The artifact is pending deployment.
     PENDING = 1;
```
Comment (alekseysidorov) [inline exonum/src/proto/schema/exonum/runtime/lifecycle.proto]: UNLOADING and PENDING looks a little bit inconsistend together. I consider that we should rename PENDING to DEPLOYING or LOADING.

**Pull request merged**